### PR TITLE
Add Thread Network Directory Cluster definition

### PR DESCRIPTION
**Pull request by ksperling-apple**

### CODE REVIEW

File: src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+<configurator>
+    <domain name="CHIP"/>
+
+    <struct name="ThreadNetworkStruct">
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml:20]: This does not match the type name in the spec, but presumably the spec is planned to be fixed here?

File: src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        <cluster code="0x0453"/>
+        <item name="ExtendedPanId" type="int64u"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml:22]: Spec has ExtendedPanID.

File: src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        <item name="NetworkName" type="char_string" length="16"/>
+        <item name="Channel" type="int16u"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml:24]: Spec has uint8.

File: src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            <access op="read" privilege="manage"/>
+            <access op="write" privilege="manage"/>
+        </attribute>
+        <attribute side="server" code="0x0001" define="THREAD_NETWORKS" type="array" entryType="ThreadNetworkStruct" writable="false" optional="false">
+            <description>ThreadNetworks</description>
+            <access op="read" privilege="operate"/>
+        </attribute>
+        <attribute side="server" code="0x0002" define="THREAD_NETWORK_TABLE_SIZE" type="int8u" writable="false" optional="false">ThreadNetworkTableSize</attribute>
+
+        <command source="client" code="0x00" name="AddNetwork" optional="false">
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml:51]: This should have `mustUseTimedInvoke="true"`, right?  Same for the other commands.

File: src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            <arg name="ExtendedPanId" type="int64u"/>
+        </command>
+        <command source="server" code="0x03" name="OperationalDatasetResponse" optional="false">
+            <description>This is the response to a GetOperationalDataset request.</description>
+            <arg name="OperationalDataset" type="octet_string" length="254"/>
+        </command>
+
+        <event side="server" code="0x00" name="NetworkChanged" priority="info" optional="false">
+            <description>This event SHALL be generated when an entry in ThreadNetworks is added, removed, or had its Operational Dataset changed.</description>
+            <field id="0" name="ExtendedPanId" type="int64u"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml:73]: Spec has ExtendedPanID.

File: src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+        <command source="client" code="0x00" name="AddNetwork" optional="false">
+            <description>Adds an entry to the ThreadNetworks list.</description>
+            <access op="invoke" privilege="manage"/>
+            <arg name="OperationalDataset" type="octet_string" length="254"/>
+        </command>
+        <command source="client" code="0x01" name="RemoveNetwork" optional="false">
+            <description>Removes an entry from the ThreadNetworks list.</description>
+            <access op="invoke" privilege="manage"/>
+            <arg name="ExtendedPanId" type="int64u"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml:59]: Spec has ExtendedPanID.

File: src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        </command>
+        <command source="client" code="0x02" name="GetOperationalDataset" optional="false" response="OperationalDatasetResponse">
+            <description>Retrieves a Thread Operational Dataset from the ThreadNetworks list.</description>
+            <access op="invoke" privilege="operate"/>
+            <arg name="ExtendedPanId" type="int64u"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml:64]: Spec has ExtendedPanID.

File: src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml (hunk 3 above)
Comment (ksperling-apple) [reply src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml:24]: I think that’s a bug in the spec, it’s uint16 in the network commissioning cluster

Comment (bzbarsky-apple) [reply src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml:24]: File spec issues/PRs as needed, please.

File: src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml (hunk 1 above)
Comment (ksperling-apple) [reply src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml:20]: Fixing in https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/9279

File: src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml (hunk 3 above)
Comment (ksperling-apple) [reply src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml:24]: Fixing in https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/9279

File: src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml (hunk 4 above)
Comment (ksperling-apple) [reply src/app/zap-templates/zcl/data-model/chip/thread-network-directory-cluster.xml]: Done

**Review by bzbarsky-apple — APPROVED** [review]

**Review by fessehaeve — APPROVED** [review]

**Pull request merged**